src/roadmap/common.py

```diff
@@ -87,54 +89,183 @@ async def query_rbac(
 
     if minor is not None:
         query = f"{query} AND system_profile_facts #>> '{{operating_system,minor}}' = :minor"
 
+    # If host group data is given, we need to filter out hosts that this user
+    # is not permitted to see. To do this we add more WHERE clauses to our
+    # query.
+    if host_groups:
+        # the hosts database keeps groups data in a JSONB field, with contents
+        # like this:
+        # [
+        #    {
+        #     "account": "123456",
+        #     "created": "2025-01-07T12:58:59.569065+00:00",
+        #     "host_count": 1,
+        #     "id": "f770fbf4-359d-11f0-b21b-5e43c8b8aa2f",
+        #     "name": "GroupTwo",
+        #     "org_id": "1234",
+        #     "ungrouped": false,
+        #     "updated": "2025-01-07T12:59:52.471612+00:00"
+        #    }
+        # ]
+
+        # The following two lines of SQL efficiently search for a match on
+        # criteria, and return TRUE if a match is found, FALSE otherwise. Here
+        # is how the lines work:
+        #
+        # * "jsonb_array_elements" queries into the denormalized JSON present
+        #   in the "groups" field. In each line the code tests a condition on a
+        #   certain field of that json data. The first line is searching for an
+        #   "ungrouped" field to have a value of "true", and the second line is
+        #   searching for the value held in the "id" field to be present in a
+        #   given set of ids.
+        # * "SELECT 1" causes the query to stop at the first match it
+        #   finds.
+        # * "EXISTS" returns a BOOLEAN instead of the result of the query.
+
+        # There is a special case. If None is in host_groups, we must query
+        # for a group that has ungrouped == True.
+        ungrouped_query = """
+            EXISTS (
+                SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj
+                    WHERE (group_obj->>'ungrouped')::boolean = true)
+        """
+        # This query searches for any group record with an id that matches any
+        # of our eligible host group ids.
+        grouped_query = """
+            EXISTS (
+                SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj
+                    WHERE group_obj->>'id' = ANY(:host_groups))
+        """
+
+        # Here we add our "group" subqueries to the WHERE clause of our final
+        # query. If the query contains both a None and string-based group id
+        # then the clauses that detect them must be comibined with on OR
+        # statement.
+        if None in host_groups:
+            if len(host_groups) > 1:
+                # Accept either a group id match or ungrouped = true.
+                query = f"{query} AND ({ungrouped_query} OR {grouped_query})"
+            else:
+                query = f"{query} AND {ungrouped_query}"
+        else:
+            query = f"{query} AND {grouped_query}"
```

Comment:
I find this section to be the most troublesome to reason about. This doesn't lower the overall complexity but I think it makes it a bit easier to read.
```suggestion
        query = f"{query} AND {grouped_query}"
        if None in host_groups:
            query = f"{query} AND {ungrouped_query}"
            if len(host_groups) > 1:
                # Accept either a group id match or ungrouped = true.
                query = f"{query} OR {grouped_query})"

```